File: shared_tests/cpp/test_runner.cpp
#include "json_loader.hpp"
#include <iostream>
#include <string>
#include <vector>
#include <filesystem>

using namespace gafro_test;

void printUsage(const char* program_name) {
    std::cout << "Usage: " << program_name << " [options] <test_file.json>\n";
    std::cout << "Options:\n";
    std::cout << "  -v, --verbose     Enable verbose output\n";
    std::cout << "  -t, --tag <tag>   Run only tests with specified tag\n";
    std::cout << "  -c, --category <name>  Run only tests in specified category\n";
    std::cout << "  -s, --stats       Show detailed statistics\n";
    std::cout << "  -r, --real-code   Enable real GAFRO code execution (Phase 2)\n";
    std::cout << "  -h, --help        Show this help message\n";
    std::cout << "\nExamples:\n";
    std::cout << "  " << program_name << " scalar_tests.json\n";
    std::cout << "  " << program_name << " -v -t basic vector_tests.json\n";
    std::cout << "  " << program_name << " -c vector_creation vector_tests.json\n";
    std::cout << "  " << program_name << " -r scalar_tests.json  # Real GAFRO execution\n";
}

void printTestSuiteInfo(const TestSuite& test_suite) {
    std::cout << "\n=== Test Suite Information ===\n";
    std::cout << "Name: " << test_suite.test_suite_name << "\n";
    std::cout << "Version: " << test_suite.version << "\n";
    std::cout << "Description: " << test_suite.description << "\n";
    
    auto stats = test_suite.getStatistics();
    std::cout << "Total Categories: " << stats.total_categories << "\n";
    std::cout << "Total Test Cases: " << stats.total_test_cases << "\n";
    
    std::cout << "\nCategories:\n";
    for (const auto& [name, count] : stats.tests_per_category) {
        std::cout << "  " << name << ": " << count << " tests\n";
    }
    
    if (!stats.tests_per_tag.empty()) {
        std::cout << "\nTags:\n";
        for (const auto& [tag, count] : stats.tests_per_tag) {
            std::cout << "  " << tag << ": " << count << " tests\n";
        }
    }
    std::cout << "==============================\n\n";
}

void printTestResults(const std::vector<TestResult>& results, bool show_stats) {
    std::cout << "\n=== Test Results ===\n";
    
    int passed = 0, failed = 0;
    double total_time = 0.0;
    
    for (const auto& result : results) {
        std::cout << "[" << (result.passed ? "PASS" : "FAIL") << "] " 
                  << result.test_name;
        
        if (show_stats) {
            std::cout << " (" << result.execution_time_ms << "ms)";
        }
        std::cout << "\n";
        
        if (result.passed) {
            passed++;
        } else {
            failed++;
            std::cout << "  Error: " << result.error_message << "\n";
        }
        
        total_time += result.execution_time_ms;
    }
    
    std::cout << "\nSummary:\n";
    std::cout << "  Passed: " << passed << "\n";
    std::cout << "  Failed: " << failed << "\n";
    std::cout << "  Total: " << (passed + failed) << "\n";
    std::cout << "  Total Time: " << total_time << "ms\n";
    
    if (passed + failed > 0) {
        std::cout << "  Average Time: " << (total_time / (passed + failed)) << "ms\n";
    }
    
    std::cout << "===================\n";
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printUsage(argv[0]);
        return 1;
    }
    
    // Parse command line arguments
    bool verbose = false;
    bool show_stats = false;
    bool real_code_execution = false;
    std::string filter_tag;
    std::string filter_category;
    std::string test_file;
    
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        
        if (arg == "-h" || arg == "--help") {
            printUsage(argv[0]);
            return 0;
        } else if (arg == "-v" || arg == "--verbose") {
            verbose = true;
        } else if (arg == "-s" || arg == "--stats") {
            show_stats = true;
        } else if (arg == "-r" || arg == "--real-code") {
            real_code_execution = true;
        } else if (arg == "-t" || arg == "--tag") {
            if (i + 1 < argc) {
                filter_tag = argv[++i];
            } else {
                std::cerr << "Error: --tag requires a tag name\n";
                return 1;
            }
        } else if (arg == "-c" || arg == "--category") {
            if (i + 1 < argc) {
                filter_category = argv[++i];
            } else {
                std::cerr << "Error: --category requires a category name\n";
                return 1;
            }
        } else if (arg[0] != '-') {
            test_file = arg;
        } else {
            std::cerr << "Error: Unknown option " << arg << "\n";
            printUsage(argv[0]);
            return 1;
        }
    }
    
    if (test_file.empty()) {
        std::cerr << "Error: No test file specified\n";
        printUsage(argv[0]);
        return 1;
    }
    
    // Check if file exists
    if (!std::filesystem::exists(test_file)) {
        std::cerr << "Error: Test file " << test_file << " does not exist\n";
        return 1;
    }
    
    // Load test suite
    std::cout << "Loading test suite from: " << test_file << "\n";
    auto test_suite = TestSuite::loadFromFile(test_file);
    
    if (!test_suite) {
        std::cerr << "Error: Failed to load test suite\n";
        return 1;
    }
    
    if (!test_suite->isValid()) {
        std::cerr << "Error: Invalid test suite\n";
        return 1;
    }
    
    // Print test suite information
    printTestSuiteInfo(*test_suite);
    
    // Set up test execution context
    TestExecutionContext context;
    context.setVerbose(verbose);
    
    // Enable real code execution if requested
    if (real_code_execution) {
        context.enableRealCodeExecution(true);
        if (verbose) {
            std::cout << "🚀 Real GAFRO code execution enabled (Phase 2)\n";
        }
    }
    
    // Execute tests based on filters
    std::vector<TestResult> results;
    
    if (!filter_category.empty()) {
        // Run specific category
        auto category = test_suite->getCategory(filter_category);
        if (!category) {
            std::cerr << "Error: Category '" << filter_category << "' not found\n";
            return 1;
        }
        
        if (!filter_tag.empty()) {
            // Filter by tag within category
            auto test_cases = category->getTestCasesByTag(filter_tag);
            for (const auto& test_case : test_cases) {
                results.push_back(context.executeTestCase(test_case));
            }
        } else {
            // Run all tests in category
            results = context.executeCategory(*category);
        }
    } else if (!filter_tag.empty()) {
        // Run all tests with specific tag
        auto test_cases = test_suite->getTestCasesByTag(filter_tag);
        for (const auto& test_case : test_cases) {
            results.push_back(context.executeTestCase(test_case));
        }
    } else {
        // Run all tests
        results = context.executeTestSuite(*test_suite);
    }
    
    // Print results
    printTestResults(results, show_stats);
    
    // Return exit code based on results
    bool all_passed = std::all_of(results.begin(), results.end(), 
                                 [](const TestResult& r) { return r.passed; });
    
    return all_passed ? 0 : 1;
}

File: shared_tests/rust/src/test_runner.rs
use clap::{Parser, ValueEnum};
use std::path::Path;
use crate::json_loader::*;

#[derive(Parser)]
#[command(name = "gafro_test_runner")]
#[command(about = "A test runner for GAFRO JSON test specifications")]
#[command(version)]
pub struct Args {
    /// Test file to run
    pub test_file: String,
    
    /// Enable verbose output
    #[arg(short, long)]
    pub verbose: bool,
    
    /// Run only tests with specified tag
    #[arg(short, long)]
    pub tag: Option<String>,
    
    /// Run only tests in specified category
    #[arg(short, long)]
    pub category: Option<String>,
    
    /// Show detailed statistics
    #[arg(short, long)]
    pub stats: bool,
    
    /// Output format
    #[arg(short, long, value_enum, default_value = "text")]
    pub format: OutputFormat,
}

#[derive(Clone, ValueEnum)]
pub enum OutputFormat {
    Text,
    Json,
}

impl std::fmt::Display for OutputFormat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            OutputFormat::Text => write!(f, "text"),
            OutputFormat::Json => write!(f, "json"),
        }
    }
}

pub fn print_usage() {
    println!("Usage: gafro_test_runner [options] <test_file.json>");
    println!("Options:");
    println!("  -v, --verbose     Enable verbose output");
    println!("  -t, --tag <tag>   Run only tests with specified tag");
    println!("  -c, --category <name>  Run only tests in specified category");
    println!("  -s, --stats       Show detailed statistics");
    println!("  -f, --format <format>  Output format (text, json)");
    println!("  -h, --help        Show this help message");
    println!();
    println!("Examples:");
    println!("  gafro_test_runner scalar_tests.json");
    println!("  gafro_test_runner -v -t basic vector_tests.json");
    println!("  gafro_test_runner -c vector_creation vector_tests.json");
}

pub fn print_test_suite_info(test_suite: &TestSuite) {
    println!("\n=== Test Suite Information ===");
    println!("Name: {}", test_suite.test_suite_name);
    println!("Version: {}", test_suite.version);
    println!("Description: {}", test_suite.description);
    
    let stats = test_suite.get_statistics();
    println!("Total Categories: {}", stats.total_categories);
    println!("Total Test Cases: {}", stats.total_test_cases);
    
    println!("\nCategories:");
    for (name, count) in &stats.tests_per_category {
        println!("  {}: {} tests", name, count);
    }
    
    if !stats.tests_per_tag.is_empty() {
        println!("\nTags:");
        for (tag, count) in &stats.tests_per_tag {
            println!("  {}: {} tests", tag, count);
        }
    }
    println!("==============================");
}

pub fn print_test_results(results: &[TestResult], show_stats: bool, format: &OutputFormat) {
    match format {
        OutputFormat::Text => print_test_results_text(results, show_stats),
        OutputFormat::Json => print_test_results_json(results, show_stats),
    }
}

fn print_test_results_text(results: &[TestResult], show_stats: bool) {
    println!("\n=== Test Results ===");
    
    let mut passed = 0;
    let mut failed = 0;
    let mut total_time = 0.0;
    
    for result in results {
        print!("[{}] {}", 
            if result.passed { "PASS" } else { "FAIL" }, 
            result.test_name
        );
        
        if show_stats {
            print!(" ({:.2}ms)", result.execution_time_ms);
        }
        println!();
        
        if result.passed {
            passed += 1;
        } else {
            failed += 1;
            println!("  Error: {}", result.error_message);
        }
        
        total_time += result.execution_time_ms;
    }
    
    println!("\nSummary:");
    println!("  Passed: {}", passed);
    println!("  Failed: {}", failed);
    println!("  Total: {}", passed + failed);
    println!("  Total Time: {:.2}ms", total_time);
    
    if passed + failed > 0 {
        println!("  Average Time: {:.2}ms", total_time / (passed + failed) as f64);
    }
    
    println!("===================");
}

fn print_test_results_json(results: &[TestResult], _show_stats: bool) {
    let mut output = serde_json::Map::new();
    
    let mut passed = 0;
    let mut failed = 0;
    let mut total_time = 0.0;
    
    let mut test_results = Vec::new();
    for result in results {
        test_results.push(JsonLoader::test_result_to_json(result));
        
        if result.passed {
            passed += 1;
        } else {
            failed += 1;
        }
        total_time += result.execution_time_ms;
    }
    
    output.insert("test_results".to_string(), serde_json::Value::Array(test_results));
    output.insert("summary".to_string(), serde_json::json!({
        "passed": passed,
        "failed": failed,
        "total": passed + failed,
        "total_time_ms": total_time,
        "average_time_ms": if passed + failed > 0 { total_time / (passed + failed) as f64 } else { 0.0 }
    }));
    
    println!("{}", serde_json::to_string_pretty(&serde_json::Value::Object(output)).unwrap_or_default());
}

pub fn run_tests(args: Args) -> Result<i32, Box<dyn std::error::Error>> {
    // Check if file exists
    if !Path::new(&args.test_file).exists() {
        eprintln!("Error: Test file {} does not exist", args.test_file);
        return Ok(1);
    }
    
    // Load test suite
    println!("Loading test suite from: {}", args.test_file);
    let test_suite = TestSuite::load_from_file(&args.test_file)?;
    
    if !test_suite.is_valid() {
        eprintln!("Error: Invalid test suite");
        return Ok(1);
    }
    
    // Print test suite information
    print_test_suite_info(&test_suite);
    
    // Set up test execution context
    let mut context = TestExecutionContext::new();
    context.set_verbose(args.verbose);
    
    // Execute tests based on filters
    let results = if let Some(category_name) = &args.category {
        // Run specific category
        if let Some(category) = test_suite.get_category(category_name) {
            if let Some(tag) = &args.tag {
                // Filter by tag within category
                let test_cases = category.get_test_cases_by_tag(tag);
                let mut results = Vec::new();
                for test_case in test_cases {
                    results.push(context.execute_test_case(&test_case));
                }
                results
            } else {
                // Run all tests in category
                context.execute_category(category)
            }
        } else {
            eprintln!("Error: Category '{}' not found", category_name);
            return Ok(1);
        }
    } else if let Some(tag) = &args.tag {
        // Run all tests with specific tag
        let test_cases = test_suite.get_test_cases_by_tag(tag);
        let mut results = Vec::new();
        for test_case in test_cases {
            results.push(context.execute_test_case(&test_case));
        }
        results
    } else {
        // Run all tests
        context.execute_test_suite(&test_suite)
    };
    
    // Print results
    print_test_results(&results, args.stats, &args.format);
    
    // Return exit code based on results
    let all_passed = results.iter().all(|r| r.passed);
    Ok(if all_passed { 0 } else { 1 })
}
